src/lib.rs

```diff
@@ -129,54 +129,54 @@ where
     }
     if let Some(build_id) = object.build_id()? {
         let path = locate_debug_build_id(build_id);
-        if path.is_ok() {
-            return path;
+        if let Some(path) = path {
+            return Ok(Some(path));
         }
         // If not found, try gnu_debuglink.
     }
     if let Some((filename, crc)) = object.gnu_debuglink()? {
         let filename = path_from_bytes(filename)?;
         return locate_gnu_debuglink(path.as_ref(), filename, crc);
     }
-    Err(anyhow!("Object does not have debug info pointer"))
+    Ok(None)
 }
 
 /// Attempt to locate the Mach-O file contained within a dSYM bundle containing the debug
 /// symbols for the Mach-O file at `path` with UUID `uuid`.
-pub fn locate_dsym<T>(path: T, uuid: Uuid) -> Result<PathBuf, Error>
+pub fn locate_dsym<T>(path: T, uuid: Uuid) -> Result<Option<PathBuf>, Error>
 where
     T: AsRef<Path>,
 {
     if let Some(dsym_path) = locate_dsym_fastpath(path.as_ref(), uuid) {
-        return Ok(dsym_path);
+        return Ok(Some(dsym_path));
     }
     locate_dsym_using_spotlight(uuid)
 }
 
 /// Attempt to locate the separate debug symbol file for the object file at `path` with
 /// build ID `id`.
-pub fn locate_debug_build_id(id: &[u8]) -> Result<PathBuf, Error> {
+pub fn locate_debug_build_id(id: &[u8]) -> Option<PathBuf> {
     if id.len() < 2 {
-        return Err(anyhow!("Build ID is too short"));
+        return None;
     }
 
     // Try "/usr/lib/debug/.build-id/12/345678etc.debug"
     let mut f = format!("/usr/lib/debug/.build-id/{:02x}/", id[0]);
     for x in &id[1..] {
-        write!(&mut f, "{:02x}", x).ok();
+        f = format!("{}{:02x}", f, x);
```

Comment:
I try not to stress out about unnecessary allocations in code that's not particularly perf-sensitive, but FYI `String` implements `std::fmt::Write`, so you can use it with `write!` like you would a file:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=be5ac4109a0729433a885ecb08bfad46

(`format!` allocates a new `String`, so you'll allocate a string per byte here as currently written.)

Reply:
Oh, you've changed the code to remove that because you're removing the `Result` return. You can do one of two things to mitigate that otherwise:
1. Just stick `?` on the end of that `.ok()`, which would early-return if it failed.
2. Ignore any possible error, something like `let _ = write!(…);`. It's not going to fail writing to a `String` unless you OOM, in which case it's going to panic anyway.

Reply:
I like option 2 and will go with that. You are right, I was trying to remove the failure case here but just ignoring a possible error and assuming writing to a string will not fail (except for a OOM which will panic) sounds like a good idea to me.